File: app/src/main/java/com/ardev/builder/compiler/Task.java
package com.ardev.builder.compiler;

import com.ardev.builder.exception.CompilationFailedException;
import com.ardev.builder.project.Project;
import com.ardev.builder.project.api.Module;

import java.io.IOException;

/**
 *
 */
public abstract class Task<T extends Module> {

    private final T mProject;

    public Task(T project) {
        mProject = project;
        
    }

   
    protected T getModule() {
        return mProject;
    }

    
    public abstract String getName();

    
    public abstract void prepare(BuildType type) throws IOException;

    
    public abstract void run() throws IOException, CompilationFailedException;

    
    protected void clean() {

    }
}

File: app/src/main/java/com/ardev/builder/compiler/Task.kt
package com.ardev.builder.compiler

import com.ardev.builder.exception.CompilationFailure
import com.ardev.builder.project.api.Module
import java.io.IOException


abstract class Task<T : Module>(private val project: T) {

    val module: T = project
    
    abstract val name: String

    @Throws(IOException::class)
    abstract fun prepare(type: BuildType)

    @Throws(IOException::class, CompilationFailure::class)
    abstract fun run()

    fun clean() {}
        
}
